marker_api/src/diagnostic.rs

```diff
@@ -148,60 +219,167 @@ impl<'ast> DiagnosticBuilder<'ast> {
-            parts: parts.as_slice().into(),
-        };
-        cx.emit_diagnostic(&diag);
+        if self.emit_lint {
+            let parts: Vec<_> = self.parts.iter().map(DiagnosticPart::to_ffi_part).collect();
+            let span = self
+                .span
+                .as_ref()
+                .expect("always `Some`, if `DiagnosticBuilder::emit_lint` is true");
+            let diag = Diagnostic {
+                lint: self.lint,
+                msg: self.msg.as_str().into(),
+                node: self.node,
+                span,
+                parts: parts.as_slice().into(),
+            };
+            cx.emit_diagnostic(&diag);
+        }
+    }
+}
+
+impl<'ast> Drop for DiagnosticBuilder<'ast> {
+    fn drop(&mut self) {
+        with_cx(self, |cx| self.emit(cx));
+    }
+}
+
+/// Every lint emission is bound to a specific node. The node is used to
+/// determine the lint level and [`Span`] that is used for the main diagnostic
+/// message.
+///
+/// This trait is implemented for most AST nodes and node ids. When given the option,
+/// it's better to use the node directly, as the id might require some callbacks into
+/// the driver to fetch the actual node.
+//
+// FIXME(xFrednet): This trait should also be implemented for all ids that implement
+// `Into<LintEmissionId>`. However, for this we first need to add more methods to fetch
+// nodes by id, to provide a valid implementation for `emission_span`.
+//
+// The `Copy` super trait is not necessary, but it allows us to simply use `impl EmissionNode`,
+// without triggering `clippy::needless_pass_by_value`. Marker could use `&impl EmissionNode`
+// instead, but then these functions would complain about values of type `*Kind` and require the
+// user to add a reference to the value. Just using `impl EmissionNode` feels better to me.
+pub trait EmissionNode<'ast>: Debug + Copy {
+    /// The [`EmissionNodeId`] which is used to determine the lint level and
+    /// where the lint is emitted.
+    fn emission_node_id(&self) -> EmissionNodeId;
+
+    /// The [`Span`] which will be used for a lint emission, if it's not overwritten by
+    /// [`DiagnosticBuilder::set_main_span`].
+    ///
+    /// The [`AstContext`] can be used to fetch the [`Span`], if this is implemented on
+    /// an id.
+    fn emission_span(&self, _cx: &AstContext<'ast>) -> Option<Span<'ast>>;
+}
+
+macro_rules! impl_emission_node_for_node {
+    ($ty:ty$(, use $data_trait:path)?) => {
+        impl<'ast> $crate::diagnostic::EmissionNode<'ast> for $ty {
+            fn emission_node_id(&self) -> $crate::diagnostic::EmissionNodeId {
+                $(
+                    use $data_trait;
+                )*
+                self.id().into()
+            }
+
+            fn emission_span(&self, _cx: &$crate::AstContext<'ast>) -> Option<$crate::ast::Span<'ast>> {
+                $(
+                    use $data_trait;
+                )*
+                Some(self.span().clone())
```

Comment:
Maybe if there were `Spanned { span() }` and `Identifiable { id() }` traits, then it would be possible to make this into a blanket impl instead of a macro boilerplate

Reply:
The `Spanned` is a nice idea! I'm not sure about `Identifiable`, I've been thinking of adding a `NodeId` enum, which basically wrapps all possible node ids, then we would have `id()` for the node-specific id, like `ExprId` for expressions and `Identifiable::node_id() -> NodeId` which can wrap all IDs.